Enhancements for Dynamic Endpoints

rynowak opened the pull request:
A series of incremental improvements for scenarios around dynamic endpoints:
- Make our built-in policies better behaved for dynamic endpoints
- Allow changing the size of the candidate set
- Use these new features in MVC

Commit:
Add some cool docs about IDynamicEndpointMetadata.

This will make James happy, which makes it more likely he will
look at the rest of the PR.

Commit:
Make HostMatcherPolicy implement IEndpointSelectorPolicy

Commit:
Implement IEndpointSelectorPolicy for HttpMethodMatcherPolicy

rynowak commented on src/Http/Routing/src/Matching/HostMatcherPolicy.cs:
```diff
@@ -48,6 +62,93 @@ public bool AppliesToEndpoints(IReadOnlyList<Endpoint> endpoints)
             });
         }
 
+        public Task ApplyAsync(HttpContext httpContext, EndpointSelectorContext context, CandidateSet candidates)
```
I didn't seem much reusable from the jump table implementation because the order of operations is different. It seemed more maintainable to just create a second code path that's about 85% separate.

rynowak commented on src/Http/Routing/src/Matching/HttpMethodMatcherPolicy.cs:
```diff
@@ -62,6 +83,104 @@ public bool AppliesToEndpoints(IReadOnlyList<Endpoint> endpoints)
+            // We want to return a 405 iff we eliminated ALL of the currently valid endpoints due to HTTP method
+            // mismatch.
+            bool? needs405Endpoint = null;
+            HashSet<string> methods = null;
```
This implemetatnion is pretty trivial without the support for 405 :(

davidfowl commented on src/Http/Routing/src/Matching/HostMatcherPolicy.cs:
```diff
@@ -48,6 +62,93 @@ public bool AppliesToEndpoints(IReadOnlyList<Endpoint> endpoints)
 
+        public Task ApplyAsync(HttpContext httpContext, EndpointSelectorContext context, CandidateSet candidates)
+        {
+            if (httpContext == null)
```
@benaadams Should this error handling be moved to another method?

rynowak replied:
This is interface dispatch. it's not going to get inlined

davidfowl replied:
Sure but there's code bloat here and nobody is going to pass null except for a unit test 😄

rynowak replied:
Don't blame me. C# should have a feature for this.

Commit:
Implement IEndpointSelectorPolicy for ConsumesMatcherPolicy

Commit:
Slim down CandidateSet

rynowak commented on src/Http/Routing/src/Matching/CandidateSet.cs:
```diff
@@ -18,19 +18,7 @@ public sealed class CandidateSet
-        private CandidateState _state2;
-        private CandidateState _state3;
-
-        private CandidateState[] _additionalCandidates;
```
@davidfowl - this boy is two allocations now but much slimmer. Some hot code paths got simpler too.

rynowak commented on src/Http/Routing/src/Matching/CandidateSet.cs:
```diff
@@ -244,15 +124,7 @@ public bool IsValidCandidate(int index)
-            {
-                return _largeCapactityValidity[index];
-            }
+            return _candidates[index].Score >= 0;
```
We're able to simplify this dramatically by using the score to store whether or not the candidate is valid

Commit:
Add ability to replace an endpoint with a collection

This is a crucial enabler for dynamic scenarios. A policy can replace an
endpoint with a *group* of dynamic endpoints which will be disambiguated
by other policies.

benaadams commented on src/Http/Routing/src/Matching/HttpMethodMatcherPolicy.cs:
```diff
@@ -62,6 +83,104 @@ public bool AppliesToEndpoints(IReadOnlyList<Endpoint> endpoints)
+                if (metadata.AcceptCorsPreflight &&
+                    string.Equals(httpMethod, PreflightHttpMethod, StringComparison.OrdinalIgnoreCase) &&
+                    httpContext.Request.Headers.ContainsKey(OriginHeader) &&
+                    httpContext.Request.Headers.TryGetValue(AccessControlRequestMethod, out var accessControlRequestMethod) &&
```
`Request.Headers.TryGetValue(AccessControlRequestMethod` 
=> `Request.Headers.TryGetValue(HeaderNames.AccessControlRequestMethod`

benaadams replied:
Also lift Request.Headers to a local rather than looking it up via 2 interfaces twice
```csharp
var headers = httpContext.Request.Headers;
if (metadata.AcceptCorsPreflight &&
    string.Equals(httpMethod, PreflightHttpMethod, StringComparison.OrdinalIgnoreCase) &&
    headers.ContainsKey(HeaderNames.OriginHeader) &&
    headers.TryGetValue(HeaderNames.AccessControlRequestMethod, out var accessControlRequestMethod) &&
    !StringValues.IsNullOrEmpty(accessControlRequestMethod))
```

e.g. non-faded are calls made getting item from RequestHeaders via httpContext:

![image](https://user-images.githubusercontent.com/1142958/56105979-d8b31b00-5f36-11e9-8dcb-89ada2e34471.png)


The long time spent in string.Equals, string.EqualsOridinalIgnoreCase for a 4 letter string "Host" is what lead to #9341 and its acceleration for `HeaderNames.` strings

rynowak replied:
Much <3 thanks!

benaadams commented on src/Http/Routing/src/Matching/HttpMethodMatcherPolicy.cs:
```diff
@@ -62,6 +83,104 @@ public bool AppliesToEndpoints(IReadOnlyList<Endpoint> endpoints)
+                var httpMethod = httpContext.Request.Method;
+                if (metadata.AcceptCorsPreflight &&
+                    string.Equals(httpMethod, PreflightHttpMethod, StringComparison.OrdinalIgnoreCase) &&
+                    httpContext.Request.Headers.ContainsKey(OriginHeader) &&
```
`Request.Headers.ContainsKey(OriginHeader`
=> `Request.Headers.ContainsKey(HeaderNames.OriginHeader`

Lookup may get faster https://github.com/aspnet/AspNetCore/pull/9341

rynowak commented:
@JamesNK - added the thing we talked about today.

Commit:
Allow fallback to controller to use MatcherPolicy

Commit:
Remove dynamic from page endpoints

Commit:
Address Bens cool feedback

rynowak commented on src/Http/Routing/src/Matching/HostMatcherPolicy.cs:
```diff
@@ -48,6 +62,93 @@ public bool AppliesToEndpoints(IReadOnlyList<Endpoint> endpoints)
+                for (var j = 0; j < hosts.Count; j++)
+                {
+                    var host = hosts[j].AsSpan();
+                    var port = "".AsSpan();
```
Made this into `ReadOnlySpan<char>.Empty`

Commit:
Address Jimmys cool feedback

rynowak commented:
@JamesNK - ready for your cool feedback

JamesNK commented on src/Http/Routing/src/Matching/CandidateSet.cs:
```diff
@@ -302,39 +167,197 @@ public void ReplaceEndpoint(int index, Endpoint endpoint, RouteValueDictionary v
+                        candidates[i] = original[i];
+                    }
+
+                    var buffer = endpoints.ToArray();
```
This would happen per-request right? I worry about CMSes out there that have tens of thousands of endpoints. Large blogs, news sites, etc. Is it worth having an API that allows a user to pass in a pre-sorted list of endpoints that they're cached?

rynowak replied:
I don't think passing in 1000 endpoints here is viable whether we pre-sort them or not. You're asking us to iterate 1000 endpoints multiple times and process policies on all of them. At that point you're holding it wrong. 

I don't think this is a implementation strategy we should try to enable. If you're writing a CMS with 1000s of endpoints you need to apply some kind of lookup yourself and not *just* use `IEndpointSelectorPolicy`. I haven't heard anyone ask for this.

JamesNK commented on src/Http/Routing/src/Matching/HttpMethodMatcherPolicy.cs:
```diff
@@ -62,6 +80,105 @@ public bool AppliesToEndpoints(IReadOnlyList<Endpoint> endpoints)
+            //
+            // Those code only runs in the presence of dynamic endpoints anyway.
+            //
+            // We want to return a 405 iff we eliminated ALL of the currently valid endpoints due to HTTP method
```
```suggestion
            // We want to return a 405 if we eliminated ALL of the currently valid endpoints due to HTTP method
```

rynowak replied:
This is intended

JamesNK commented on src/Mvc/Mvc.Core/src/Routing/ConsumesMatcherPolicy.cs:
```diff
@@ -24,16 +22,140 @@ internal class ConsumesMatcherPolicy : MatcherPolicy, IEndpointComparerPolicy, I
+                throw new ArgumentNullException(nameof(candidates));
+            }
+
+            // We want to return a 415 iff we eliminated ALL of the currently valid endpoints due to content type
```
```suggestion
            // We want to return a 415 if we eliminated ALL of the currently valid endpoints due to content type
```

rynowak replied:
No, this is correct. If and only if.

rynowak commented:
@aspnet-hello triage https://dev.azure.com/dnceng/public/_build/results?buildId=157692&view=logs

rynowak commented:
@jamesnk any other feedback?

aspnet-hello commented:
This comment was made automatically. If there is a problem contact ryanbrandenburg.

I've triaged the above build.
